### ISSUE TITLE
401 with "new-style" credential provider plugin

### ISSUE THREAD

**Issue by quasarea** [microsoft/artifacts-credprovider#137]
We switched to explicitly use new method, nuget is no more in same location as the plugin and plugin is pointed by NUGET_PLUGIN_PATHS env variable.

Nuget 5.2.0.6090
Command-line v0.1.19

This time we can get:
```
..\..\Build\NuGet\NuGet.exe install MyPackage -NonInteractive -ConfigFile ..\..\Build\NuGet\NuGet.Config -Prerelease -Framework net472 -ExcludeVersion -DependencyVersion Lowest -OutputDirectory ..\..\Build\NuGet\Tools -Version 2.4.6.30
Feeds used:
  https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json


Attempting to gather dependency information for package 'MyPackage.2.4.6.30' with respect to project '..\..\Build\NuGet\Tools', targeting '.NETFramework,Version=v4.7.2'
Problem starting the plugin 'M:\Repo\Build\NuGet\CredentialProvider\CredentialProvider.Microsoft.exe'. Plugin 'CredentialProvider.Microsoft' failed within 6.279 seconds with exit code -1.
Problem starting the plugin 'M:\Repo\Build\NuGet\CredentialProvider\CredentialProvider.Microsoft.exe'. Plugin 'CredentialProvider.Microsoft' failed within 6.279 seconds with exit code -1.
  Plugin 'CredentialProvider.Microsoft' failed within 6.279 seconds with exit code -1.
  A task was canceled.
An error occurred while retrieving package metadata for 'MyPackage.2.4.6.30' from source 'CompanyFeed'.
  Response status code does not indicate success: 401 (Unauthorized).
```

And internal log 
```
[18:38:05.441 11584 Verbose] Running in plug-in mode
[18:38:05.448 11584 Verbose] Command-line v0.1.19: "M:\Repo\Build\NuGet\CredentialProvider\CredentialProvider.Microsoft.exe" -Plugin
[18:38:05.519 11584 Verbose] Handling 'Request' 'Initialize'. Time elapsed in ms: 22 - Payload: {"ClientVersion":"5.2.0","Culture":"en-GB","RequestTimeout":"00:00:05"}
[18:38:05.520 11584 Verbose] Sending response: 'Request' 'Initialize'. Time elapsed in ms: 24
[18:38:05.522 11584 Verbose] Time elapsed in milliseconds after sending response 'Request' 'Initialize': 25
[18:38:05.531 11584 Verbose] Handling 'Request' 'GetOperationClaims'. Time elapsed in ms: 1 - Payload: {}
[18:38:05.532 11584 Verbose] Sending response: 'Request' 'GetOperationClaims'. Time elapsed in ms: 2
[18:38:05.556 11584 Verbose] Time elapsed in milliseconds after sending response 'Request' 'GetOperationClaims': 26
[18:38:05.617 11584 Verbose] Handling 'Request' 'SetLogLevel'. Time elapsed in ms: 2 - Payload: {"LogLevel":"Information"}
[18:38:05.619 11584 Verbose] Sending response: 'Request' 'SetLogLevel'. Time elapsed in ms: 3
[18:38:05.620 11584 Verbose] Time elapsed in milliseconds after sending response 'Request' 'SetLogLevel': 5
[18:38:05.640 11584 Verbose] Handling 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 5 - Payload: {"Uri":"https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json","IsRetry":false,"IsNonInteractive":true,"CanShowDialog":true}
[18:38:05.641 11584 Verbose] Creating a progress reporter with interval: 00:00:02
[18:38:05.647 11584 Verbose] Handling auth request, Uri: https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json, IsRetry: False, IsNonInteractive: True, CanShowDialog: True
[18:38:05.647 11584 Verbose] URI: https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json
[18:38:05.649 11584 Verbose] VstsBuildTaskServiceEndpointCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials. Appropriate environment variable needs to be set.
[18:38:05.649 11584 Verbose] Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint.VstsBuildTaskServiceEndpointCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json
[18:38:05.650 11584 Verbose] VstsBuildTaskCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet. Appropriate environment variables must be set.
[18:38:05.651 11584 Verbose] Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTask.VstsBuildTaskCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json
[18:38:06.436 11584 Verbose] GET https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json
[18:38:06.746 11584 Verbose] IsRetry: False
[18:38:06.775 11584 Verbose] Found cached SessionToken for https://pkgs.dev.azure.com/Company/_packaging/CompanyFeed/nuget/v3/index.json
[18:38:06.776 11584 Verbose] Sending response: 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 1142
[18:38:06.782 11584 Verbose] Time elapsed in milliseconds after sending response 'Request' 'GetAuthenticationCredentials': 1147
```

**Comment by tkrick** [comment]
this looks like the cancellation issue that we are working with the nuget team on. [link](https://github.com/microsoft/artifacts-credprovider/issues/91#issuecomment-530907753). can you try some of those steps? I see startup is taking longer than 5 seconds so it is probably that case. thanks.

**Comment by SaintGimp** [comment]
For what it's worth, I'm seeing the same issue with NuGet 5.3.1.6268.  On some of my (self-hosted) build agents, it's picking up the V1 credential manager and succeeds.  On other agents it's picking up the V2 credential manger and is failing with a 401 error.  I'm not entirely sure what the difference is between the agents, but it's very consistent.

**Comment by SaintGimp** [comment]
Ok, I think I have a fix for my specific situation.  My pipeline specified the NugetCommand task with a externalFeedCredentials parameter which pointed to a valid service connection with a valid PAT.  As noted above, some of my agents were using the V1 credential manager and some were using V2.  At least, I assume V1 was being invoked because there wasn't any logging output about the credential manager at all.  I still don't know why this difference in behavior existed, but the V1 credential manager was able to access the service connection correctly but the V2 credential manager failed to access the service connection.  Specifically, V2 failed here:
```
[CredentialProvider]Sending response: 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 0
[CredentialProvider]Time elapsed in milliseconds after sending response 'Request' 'GetAuthenticationCredentials': 0
NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://redacted.pkgs.visualstudio.com/_packaging/redacted/nuget/v3/index.json. ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
```
I modified the NugetCommand task and removed the externalFeedCredentials parameter entirely.  Now the pipeline invokes the V2 credential manager on all agents and successfully authenticates to the feed.

**Comment by pspill** [comment]
This is NuGet/Home#8528

**Comment by satbai** [comment]
Should be fixed with NuGet [5.5.x](https://www.nuget.org/downloads) and cred provider [0.1.20](https://github.com/microsoft/artifacts-credprovider/releases)
